### PR TITLE
RF: Use the average sigma in the mask.

### PR DESCRIPTION
**Pull request by arokem**
Instead of the value of sigma in the corner of the image.

Addresses #1131

### CODE REVIEW

**Comment by arokem** [comment]
Urgh. What's the right way to handle `nanmean`? It's not in `scipy.stats` in newer versions of scipy, and it's not in `numpy` in older versions of numpy. Do I really have to check the version of both numpy and scipy to be sure that I can compute a nanmean?

**Comment by MarcCote** [comment]
Maybe compute it explicitly instead of relying on nanmean? See http://stackoverflow.com/questions/5480694/numpy-calculate-averages-with-nans-removed for some variants.

NB: Maybe a bit slower compared to np.nanmean, though.

**Comment by arokem** [comment]
Thanks! That looks like a good solution to me. We can check whether numpy has a nanmean and use that one if possible, and otherwise use our own. What do you think?

**Comment by MarcCote** [comment]
I liked it. :+1:

**Comment by arokem** [comment]
What do you think about all the rest of this? Does it properly address #1131 in your opinion?

**Comment by arokem** [comment]
Any more thoughts here? Seems like a bug we need to address before we can make a release.

**Comment by arokem** [comment]
Anyone want to take a look at this?

**Comment by MarcCote** [comment]
This looks good to me. @samuelstjean last chance to have a look at this PR. Does it address your issue #1131 correctly?

**Comment by samuelstjean** [comment]
It kinds of circumvent the problem. It is a less brute approach than picking the top left corner voxel, but it does cause problem if you have a large variation of the noise profile in your image. In that case, I would prefer if it just threw me an error that 3D arrays are not supported, and let me choose how to deal with it. 

For example, say I collected a noise measurement maps, extract local standard deviation from that, and feed it to the function. It will pick out the mean, and if the map is first masked (say by the scanner), then the median would probably be more appropriate to deal with those 3/4 of zeros values. As of now, I'll probably get a useless value, where the previous nlmeans version would happily work with my 3D map at each spatial location. Since that is now the case anymore, this would all happen silently and give me a less optimal result than I expected (true story, ask JC).

**Comment by arokem** [comment]
I am back to this. I agree with @samuelstjean. It doesn't make sense to accept 3D sigma inputs to this function, because it will create a scalar under the hood using less than obvious rules. If you want to extract your sigma from a 3D volume, please do so before you feed it to this function. Simplified accordingly.

**Comment by arokem** [comment]
Now rebased on master

**Comment by MarcCote** [comment]
@arokem @samuelstjean I'll be merging this PR on Monday.

**Comment by samuelstjean** [comment]
+ the nanmean function serves no purpose
+ not sure what error message it will throw you if you pass a 3D array, and it might be hard to decipher

not sure it is fully finished in this current form, but the logic seems there

**Comment by arokem** [comment]
Good comments both -- thanks for taking a look. I've removed the nanmean
cruft and added a check up-front for scalar sigma.

**Comment by arokem** [comment]
I added the option of passing in an array, if that array is a singleton. This fixes the test failures from yesterday (at least on my machine), and actually seems reasonable to me. I don't think that we need to adjust the documentation. The canonical input is still a single float, it's just that if you happen to deliver that single float in an array, we'll still accept that. 

Now also rebased on master

File: dipy/denoise/tests/test_non_local_means.py [hunk 1]
```diff
@@ -81,7 +96,7 @@ def test_nlmeans_dtype():
     S0 = 200 * np.ones((20, 20, 20), dtype=np.uint16)
     mask = np.zeros((20, 20, 20))
     mask[10:14, 10:14, 10:14] = 1
-    S0n = non_local_means(S0, sigma=np.ones((20, 20, 20)), mask=mask,
+    S0n = non_local_means(S0, sigma=1, mask=mask,
```
Comment (MarcCote) [inline dipy/denoise/tests/test_non_local_means.py]: That probably fits on one line now.

File: dipy/denoise/tests/test_non_local_means.py [hunk 2]
```diff
@@ -39,7 +55,7 @@ def test_nlmeans_boundary():
 
     S0[:10, :10, :10] = 300 + noise[:10, :10, :10]
 
-    S0n = non_local_means(S0, sigma=np.ones((20, 20, 20)) * np.std(noise),
+    S0n = non_local_means(S0, sigma=np.std(noise),
```
Comment (MarcCote) [inline dipy/denoise/tests/test_non_local_means.py]: That probably fits on one line now.

**Comment by arokem** [comment]
Indeed. I've removed some other un-needed whitespace, while I was at it.

**Pull request merged**